### ISSUE TITLE
Calling layout on TextPainter is slow, and gets progressively slower on MacOS and Web 

### ISSUE THREAD

**Issue by Dall127** [flutter/flutter#132421]
### To recreate: 

Run TextPainter.layout multiple times, and it gets progressively worse; My main program runs layout a lot to conditionally paint text on a page, and it slowly crawls to a stop. I've provided sample code, and other requested information

- I ran it on both flutter stable branch and flutter master, both producing the same issue
-  ``` flutter run --profile --trace-skia ``` produced no output
- flutter analyze produces no issues

```

import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Painting demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int difference = 0;

  StringBuffer buffer = StringBuffer("Hello world");
  TextStyle style = const TextStyle(fontSize: 20, color: Colors.black);
  TextPainter painter = TextPainter(textDirection: TextDirection.ltr);

  void paintMultiple() {
    final startTime = DateTime.now();
    difference = 0;
    for (int i = 0; i < 200; i++) {
      buffer.write("Hello world");
      painter.text = TextSpan(text: buffer.toString(), style: style);
      painter.layout();
    }
    final endTime = DateTime.now();

    setState(() {
      difference = endTime.difference(startTime).inMilliseconds;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title:
              Text(widget.title, style: const TextStyle(color: Colors.white)),
          backgroundColor: Colors.purple,
        ),
        body: Center(
            child: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            if (difference > 0) Text("Time to paint: $difference milliseconds"),
            if (difference == 0) const Text("Time to paint: not painted"),
            const SizedBox(height: 20),
            OutlinedButton(
              onPressed: () => paintMultiple(),
              child: const Text("Paint multiple times"),
            )
          ],
        )));
  }
}

```

MacOS - Apple M1 Max - 32 GB Ram - Ventura version 13.5 (22G74)
![image](https://github.com/flutter/flutter/assets/31700756/c4a4bacd-d729-4e7d-a24c-33d4305bac17)

Chrome - Version 115.0.5790.170 (Official Build) (arm64)
![image](https://github.com/flutter/flutter/assets/31700756/c84ab653-86b3-46c2-99a7-27dfc6c41437)

https://github.com/flutter/flutter/assets/31700756/30bca1ec-cb1d-4304-b16f-994a030da10d

```
[redacted]@[redacted] painter_sample % flutter doctor -v
[✓] Flutter (Channel master, 3.14.0-3.0.pre.45, on macOS 13.5 22G74 darwin-arm64, locale en-US)
    • Flutter version 3.14.0-3.0.pre.45 on channel master at /Users/[redacted]/LocalFrameworks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 36c40e4972 (50 minutes ago), 2023-08-11 18:48:13 -0400
    • Engine revision 5669544866
    • Dart version 3.2.0 (build 3.2.0-62.0.dev)
    • DevTools version 2.26.1

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/[redacted]/Library/Android/sdk
    • Platform android-33, build-tools 32.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] VS Code (version 1.81.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.70.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.5 22G74 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 115.0.5790.170

[✓] Network resources
    • All expected network resources are available.

• No issues found!


```

**Comment by moffatman** [comment]
You can/should cache laid out paragraphs and not relayout them every frame if possible.

But specifically in your example, you are actually laying out the text "Hello world" a total of 20 thousand times just on the first paintMultiple. And that accumulates each time you click the button. For example, the second click would layout strings totalling 60,000 "Hello world". And the third 100,000, fourth 140,000, etc, etc. Is that intentional? (because you don't clear buffer each time)?

**Comment by Dall127** [comment]
Thanks for responding, @moffatman. The primary goal is to sequentially lay out sentences in a frame until I reach an "overflow" height, at which I then create the subsequent page, and continue in this manner. Moreover, when the display I'm working on is adjusted (e.g., when a web screen is resized or a phone is rotated), I must repeat this process for every page. I concur that invoking this repeatedly can be computationally demanding. However, I've optimized it so that it's triggered only around 10-20 times per page. I believe there's likely a more efficient caching method, but despite extensively reviewing the documentation and source code for the TextPainter class, I couldn't devise a strategy to manage and repurpose the objects it generates. From my understanding, any alteration in constraints or text necessitates a rebuild. My sample code might not have been the best illustration since the buffer wasn't cleared after a certain number of paints, but my main point was to emphasize that frequent calls can introduce significant jank, which is only exacerbated by lower-end to mid-range machines, and probably adds a significant amount of overhead on all Flutter apps that have a lot of text, or do a lot of text manipulation

**Comment by moffatman** [comment]
Not sure how are you doing this sequential layout, if you can provide a different example, it will be more helpful, because this one doesn't really demonstrate any specific issues in the text layout system, as its "progressive slowness" is because of a progressively growing workload.

There is an optimization I am thinking of proposing for the paragraph APIs that could help in your use case, where layout with a specified max-height would stop doing more work once the max height was reached. At the moment, when you use maxHeight on a paragraph of flutter text, it will do the work for all provided text, even if we know that we are always going to clip at for example 100px in height.

**Comment by Dall127** [comment]
Sorry, I think that I should've rather titled this Calling layout on TextPainter is slow, and left off the progressive bit as that was just wrong, and I'll remove that from the title.

I think the main thing is that would be beneficial to my use case is just reusing objects as much as possible; The way that I'm progressively laying out text is that I add sentence at a time, calculating if I've overflowed, and then sending it to the next page to have the same thing done, so that the text is truly 'paged'. You could think of it like reading a Kindle, per se.  I'm not super familiar with the underlying APIs, but I would've imagined that instead of creating a new paragraph object for every layout (which seems to happen when either text or styles change) that it would instead try to reuse the existing paragraph objects instead of instantiating a new one every time.

**Comment by moffatman** [comment]
Yeah, I agree, and I do some ideas for more efficent layout such as you describe. For example, make a paragraph with your full text, but give it the maximum height, and it could quit doing work one reached that height, and output how far it got. So no speculative layout would ever be needed, you could directly start the next paragraph where the first one ended.

**Comment by Dall127** [comment]
@moffatman, that approach seems promising. Put succinctly, my main concern is ensuring the correct pagination of text across various page objects, and I believe that approach would allow me to drastically cut down on my calls to layout if I knew what text content is overflowed.

**Comment by moffatman** [comment]
Something which may help you today, is to layout some reasonable amount of the text (a guess around 1.2x page height). Then use `TextPainter.getPositionForOffset(Offset(0, maxHeight + 1 px))` to find the text position that breaks to the next page. Note this means you still do extra work laying out the text which exceeded the page (which my change above could fix). But it prevents laying out from the beginning an increasingly growing paragraph, which if it takes 10 tries as you say, is 5.5x the work of getting it right the first time.

**Comment by verger-guo** [comment]
Have you found a good solution to this issue?
